### ISSUE TITLE
Typeahead has document click handler even when autocomplete not in use

### ISSUE THREAD

**Issue by deeg** [ng-bootstrap/ng-bootstrap#1471]
### Bug description:

For every typeahead you have on a page, there is a document click handler which handles dismissing a popup. This event is always bound regardless of if the input has focus, or a typeahead popup is open.

This means every time a user clicks a page, the dismissPopup function is going to run for each typeahead on the screen.

We could change to something like this which we subscribe/unsubscribe to when needed.

`
  private documentClickObservable = Observable.fromEvent(document, 'click');
  private documentClickSubscription: Subscription;
`

When you subscribe to the observable, the callback will fire only until you unsubscribe, in which case it will lie dormant.

We could subscribe when a popup is opened, and unsubscribe when it is closed.

I can put out a PR for this, I wanted to hear people's thoughts first.

### Link to minimally-working plunker that reproduces the issue:

This can be found by putting a breakpoint in the dismissPopup function in the demo, and clicking around on the page.

### Version of Angular, ng-bootstrap, and Bootstrap:

Angular: 4.0.1

ng-bootstrap: 1.0.0-alpha.22

Bootstrap: 4.0.0-alpha.6

**Comment by wesleycho** [comment]
We cannot just use raw `Observable.fromEvent` because we need to also be server-side friendly - we need to use `Renderer` (or `Renderer2` if using Angular 4+).

Maybe the best solution would be to create a service to manage typeaheads for the purpose of managing which one is dismissed.

**Comment by pkozlowski-opensource** [comment]
Yes, this is essentially the same as the issue we've got already for dropdowns: #900

PRs are welcomed, but ideally we should come up with a solution that covers all the existing popups we create (dropdown, typeahead but also datepicker popup, popover etc.).

**Comment by wesleycho** [comment]
Maybe a service called `NgbEventManager` whose purpose is to manage events bound to globals like `document`?

**Comment by pkozlowski-opensource** [comment]
So, I was discussing this recently with my colleagues and as far as I can tell we've got 2 basic solutions:
* subscribe / unsubscribe to / from global events based on the open / closed state
* insert a backdrop element _behind_ an opened popup and have this element capture click events; this is a solution similar to a backdrop being inserted for modals.

TBH I'm not sure which solution is better but whatever we do it should work for all the widgets that open popups (dropdown, datepicker, tooltip / popover, typeahead)

**Comment by wesleycho** [comment]
IMO the first solution is much better - it avoids being opinionated by introducing HTML/CSS into the mix, would could lead to problems if users have particular UX needs.

I think that Angular should provide this feature in core itself (https://github.com/angular/angular/issues/15905), but I don't think we can wait for Angular to implement this - it may be better for us to implement a version specific to our purposes and remove it in favor of Angular's version down the line.

**Comment by wKoza** [comment]
So @wesleycho , @pkozlowski-opensource , what can we do ? 

- subscribe / unsubscribe , for each widget, on the open / close state with a code like this : 

`private _openPopup() {
      this.removeListenerClickonDocument = this._renderer.listen('document', 'click', this.dismissPopup.bind(this));
}`

**Comment by wKoza** [comment]
So @wesleycho , @pkozlowski-opensource , what can we do ?

- subscribe / unsubscribe , for each widget, on the open / close state with a code like this :

```javascript
private _openPopup() { 
this.removeListenerClickonDocument = this._renderer.listen('document', 'click', this.dismissPopup.bind(this)); 
}

 private _closePopup() {
this.removeListenerClickonDocument();
}
```

- Create a service for this (but I'm confused about the fact to use the `Renderer` in a service.

?
